src/out_of_order_engine.cc

```diff
@@ -94,6 +94,9 @@ struct engine_impl {
 	/// is assumed to have completed earlier (triggering its removal from the map).
 	std::unordered_map<instruction_id, incomplete_instruction_state> incomplete_instructions;
 
+	/// The subset of incomplete_instructions that to not have incomplete predecessors themselves.
+	std::unordered_set<instruction_id> completion_front;
```

Comment:
Minor point, but I think it would make more sense to name this something like `incomplete_front`, or maybe `pending_instructions`. (I would assume that a "completion_front" is a set of complete instructions)

Reply:
Since the `out_of_order_engine` only really deals with incomplete instructions, I think it's quite clear from context. That being said, how about 'execution_front`?

Reply:
I'm ok with `execution_front`, but `incomplete_front` and `pending_instructions` are too unspecific IMO.

Reply:
I've renamed it to `execution_front` and also added some explicit tests on that.